server/protocols/webadmin/webadmin-mailqueue/src/main/java/org/apache/james/webadmin/routes/MailQueueRoutes.java

```diff
@@ -238,4 +251,107 @@ private boolean filter(MailQueueItemDTO item, Optional<Boolean> isDelayed) {
+                (request, response) -> deleteMails(request, response));
+    }
+
+    private Object deleteMails(Request request, Response response) {
+        String mailQueueName = request.params(MAIL_QUEUE_NAME);
+        Object bodyResponse = mailQueueFactory.getQueue(mailQueueName)
+            .map(name -> deleteMails(name, 
+                    sender(request.queryParams(SENDER_QUERY_PARAM)),
+                    name(request.queryParams(NAME_QUERY_PARAM)),
+                    recipient(request.queryParams(RECIPIENT_QUERY_PARAM))))
+            .orElseThrow(
+                () -> ErrorResponder.builder()
+                    .message(String.format("%s can not be found", mailQueueName))
+                    .statusCode(HttpStatus.NOT_FOUND_404)
+                    .type(ErrorResponder.ErrorType.NOT_FOUND)
+                    .haltError());
+        response.status(HttpStatus.NO_CONTENT_204);
+        return bodyResponse;
+    }
+
+    private Optional<MailAddress> sender(String senderAsString) throws HaltException {
+        try {
+            return Optional.ofNullable(senderAsString)
+                    .map(Throwing.function((String sender) -> new MailAddress(sender)).sneakyThrow());
+        } catch (Exception e) {
+            throw ErrorResponder.builder()
+                .statusCode(HttpStatus.BAD_REQUEST_400)
+                .type(ErrorType.INVALID_ARGUMENT)
+                .message("'sender' should be a MailAddress")
+                .cause(e)
+                .haltError();
+        }
+    }
+
+    private Optional<String> name(String nameAsString) {
+        return Optional.ofNullable(nameAsString);
+    }
+
+    private Optional<MailAddress> recipient(String recipientAsString) throws HaltException {
+        try {
+            return Optional.ofNullable(recipientAsString)
+                    .map(Throwing.function((String recipient) -> new MailAddress(recipient)).sneakyThrow());
+        } catch (Exception e) {
+            throw ErrorResponder.builder()
+                .statusCode(HttpStatus.BAD_REQUEST_400)
+                .type(ErrorType.INVALID_ARGUMENT)
+                .message("'recipient' should be a MailAddress")
+                .cause(e)
+                .haltError();
+        }
+    }
+
+    private Object deleteMails(ManageableMailQueue queue, Optional<MailAddress> maybeSender, Optional<String> maybeName, Optional<MailAddress> maybeRecipient) {
+        if (!maybeSender.isPresent() && !maybeName.isPresent() && !maybeRecipient.isPresent()) {
+            throw ErrorResponder.builder()
+                .statusCode(HttpStatus.BAD_REQUEST_400)
+                .type(ErrorType.INVALID_ARGUMENT)
+                .message("At least one of 'sender', 'name' or 'recipient' query parameter should be given.")
+                .haltError();
+        }
+        maybeSender.ifPresent(sender -> taskManager.submit(new MailQueueSenderDeletionTask(queue, sender)));
```

Comment:
I don't think we should run 3 deletions, it should be a single deletion with more or less filters. The more you give parameters, the more it should filter things. Unless the backend does not provide the feature and you have to reject request with more than 1 filter.

Reply:
Can not more agree with this!